extensions/standard-processors/processors/PutTCP.cpp

```diff
@@ -0,0 +1,551 @@
+  return ssl_context;
+}
+
+template<>
+nonstd::expected<std::shared_ptr<SslSocket>, std::error_code> ConnectionHandler<SslSocket>::establishConnection(const tcp::resolver::results_type& resolved_query) {
+  auto ssl_context = getSslContext(ssl_context_service_);
+  auto socket = std::make_shared<SslSocket>(io_context_, ssl_context);
+  startConnect(resolved_query.begin(), socket);
+  deadline_.async_wait([&](const std::error_code& error_code) -> void {
+    checkDeadline(error_code, socket);
+  });
+  io_context_.run();
+  if (last_error_)
+    return nonstd::make_unexpected(last_error_);
+  return socket;
+}
+
+template<class SocketType>
+nonstd::expected<void, std::error_code> ConnectionHandler<SocketType>::sendDataToSocket(const std::shared_ptr<SocketType>& socket,
+                                                                                        const std::vector<std::byte>& data,
+                                                                                        const std::vector<std::byte>& delimiter) {
+  if (!socket || !socket->lowest_layer().is_open())
+    return nonstd::make_unexpected(asio::error::not_socket);
+
+  deadline_.expires_after(timeout_);
+  io_context_.restart();
+
+  asio::async_write(*socket, asio::buffer(data), [&](const std::error_code err, std::size_t bytes_written) {
+    handleWrite(err, bytes_written, delimiter, socket);
+  });
+  deadline_.async_wait([&](const std::error_code& error_code) -> void {
+    checkDeadline(error_code, socket);
+  });
+  io_context_.run();
+  if (last_error_)
+    return nonstd::make_unexpected(last_error_);
+  last_used_ = std::chrono::steady_clock::now();
+  return {};
+}
+
+template<class SocketType>
+nonstd::expected<tcp::resolver::results_type, std::error_code> ConnectionHandler<SocketType>::resolveHostname() {
+  tcp::resolver resolver(io_context_);
+  std::error_code error_code;
+  auto resolved_query = resolver.resolve(connection_id_.getHostname(), connection_id_.getPort(), error_code);
+  if (error_code)
+    return nonstd::make_unexpected(error_code);
+  return resolved_query;
+}
+}  // namespace
+
+void PutTCP::onTrigger(core::ProcessContext* context, core::ProcessSession* const session) {
+  gsl_Expects(context && session);
+
+  const auto flow_file = session->get();
+  if (!flow_file) {
+    yield();
+    return;
+  }
+
+  removeExpiredConnections();
+
+  auto hostname = context->getProperty(Hostname, flow_file).value_or(std::string{});
+  auto port = context->getProperty(Port, flow_file).value_or(std::string{});
+  if (hostname.empty() || port.empty()) {
+    logger_->log_error("[%s] invalid target endpoint: hostname: %s, port: %s", flow_file->getUUIDStr(),
+        hostname.empty() ? "(empty)" : hostname.c_str(),
+        port.empty() ? "(empty)" : port.c_str());
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  auto data = session->readBuffer(flow_file);
```

Comment:
While this works, would it be possible to avoid reading the whole flow file into memory, and only keep the chunks in memory as they are being sent?

For example, instead of passing a vector to the handler functions, you could pass a stream, and the callbacks could read a chunk into temporary buffers.

Reply:
good idea, I've changed it (with hard coded 1024 byte chunk size)
https://github.com/apache/nifi-minifi-cpp/pull/1419/commits/ebd7db571a95de757b6a7f4c98f8a22fc1b849c8